### PR TITLE
Merklekeystore

### PR DESCRIPTION
**Pull request by algonathan**
## Summary
Merklekeystore contains a new struct to contain multiple keys.

### CODE REVIEW

File: crypto/merklekeystore/merklekeystore_test.go [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+		a.Error(signer.GetVerifier().Verify(hashable, sig))
+
+		sig2 := sig
+		someDigest := crypto.Digest{}
+		rand.Read(someDigest[:])
+		sig2.Proof[0] = someDigest
+		a.Error(signer.GetVerifier().Verify(hashable, sig))
+	})
+
+	t.Run("bad leaf position in signature", func(t *testing.T) {
```
Comment (algoidan) [inline crypto/merklekeystore/merklekeystore_test.go:147]: in order to be consistent with the rest of the code. don't use r.Run

File: crypto/merklekeystore/signer.go [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+var errStartBiggerThanEndRound = fmt.Errorf("cannot create merkleKeyStore because end round is smaller then start round")
+
+// New Generates a merklekeystore.Signer
+func New(startRound, endRound uint64) (*Signer, error) {
+	if startRound > endRound {
+		return nil, errStartBiggerThanEndRound
+	}
+	keys := make(disposableKeys, endRound-startRound)
+	for i := range keys {
+		keys[i] = crypto.NewSigner(crypto.PlaceHolderType)
```
Comment (algoidan) [inline crypto/merklekeystore/signer.go:70]: we should get the type as an arg to that function.

File: crypto/merklekeystore/signer.go [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+import (
+	"fmt"
+	"github.com/algorand/go-algorand/crypto"
+	"github.com/algorand/go-algorand/crypto/merklearray"
+)
+
+type disposableKeys []*crypto.SignatureAlgorithm
+
+//Length returns the amount of disposable keys
+func (d disposableKeys) Length() uint64 {
```
[the remark this answers is not in the document]
Comment (algonathan) [reply crypto/merklekeystore/signer.go:28]: The type itself is a rename for a slice. 
i think there shouldn't be a copy there

File: crypto/merklekeystore/signer.go [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+	disposableKeys
+	startRound uint64
+	tree       *merklearray.Tree
+}
+
+var errStartBiggerThanEndRound = fmt.Errorf("cannot create merkleKeyStore because end round is smaller then start round")
+
+// New Generates a merklekeystore.Signer
+func New(startRound, endRound uint64) (*Signer, error) {
+	if startRound > endRound {
```
[the remark this answers is not in the document]
Comment (algonathan) [reply crypto/merklekeystore/signer.go:65]: i've changed the doc to state that the key can be in [startRound, endRound). 
which conforms with things like 0,1,2,3,4,5...n-1 in code.
do you want me to change it to [startRound, endRound]?

File: crypto/merklekeystore/merklekeystore_test.go [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+	a.NoError(err)
+	a.NoError(signer.GetVerifier().Verify(hashable, sig))
+
+	_, err = signer.Sign(hashable, basics.Round(start-1))
+	a.Error(err)
+
+	_, err = signer.Sign(hashable, basics.Round(end+1))
+	a.Error(err)
+
+	TestBadLeafPositionInSignature(t)
```
Comment (algoidan) [inline crypto/merklekeystore/merklekeystore_test.go:131]: why do we want to run this test like this?

File: crypto/merklekeystore/signer.go [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+	Pos                  uint64              `codec:"pos"`
+}
+
+// Signer is a merkleKeyStore, contain multiple keys which can be used per round.
+type Signer struct {
+	_struct struct{} `codec:",omitempty,omitemptyarray"`
+	// these keys are the keys used to sign in a round.
+	// should be disposed of once possible.
+	EphemeralKeys    `codec:"keys,allocbound=-"`
+	FirstRound       basics.Round `codec:"sround"`
```
Comment (algoidan) [inline crypto/merklekeystore/signer.go:67]: srnd?

File: crypto/merklekeystore/merklekeystore_test.go [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+	sig2.Proof = sig2.Proof[:len(sig2.Proof)-1]
+	a.Error(signer.GetVerifier().Verify(hashable, sig2))
+
+	sig3 := sig2
+	someDigest := crypto.Digest{}
+	rand.Read(someDigest[:])
+	sig3.Proof[0] = someDigest
+	a.Error(signer.GetVerifier().Verify(hashable, sig3))
+}
+
```
Comment (algoidan) [inline crypto/merklekeystore/merklekeystore_test.go:170]: should we create test empty Merkle proof?

File: crypto/merklekeystore/merklekeystore_test.go [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+	signer, err := New(0, 100, crypto.PlaceHolderType)
+	a.NoError(err)
+
+	for i := 0; i < 100; i++ {
+		pos, err := signer.getKeyPosition(basics.Round(i))
+		a.NoError(err, i)
+		a.Equal(uint64(i), pos)
+	}
+
+	_, err = signer.getKeyPosition(100)
```
Comment (algoidan) [inline crypto/merklekeystore/merklekeystore_test.go:60]: should we move it into the loop?

File: crypto/merklekeystore/merklekeystore_test.go [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+		a.NoError(err)
+		_, err = signer.Sign(h, basics.Round(i))
+		a.NoError(err)
+	}
+
+	_, err := New(1, 0, crypto.PlaceHolderType)
+	a.Error(err)
+
+	signer, err := New(0, 0, crypto.PlaceHolderType)
+	a.NoError(err)
```
Comment (algoidan) [inline crypto/merklekeystore/merklekeystore_test.go:43]: maybe validate that there is 1 ephemeral key?

File: crypto/merklekeystore/merklekeystore_test.go (hunk 8 above)
Comment (algonathan) [reply crypto/merklekeystore/merklekeystore_test.go:60]: thanks for catching it, that's a bug. i'll fix it

File: protocol/hash.go [hunk 10]
```diff
@@ -55,4 +55,6 @@ const (
 	TxnMerkleLeaf     HashID = "TL"
 	Transaction       HashID = "TX"
 	Vote              HashID = "VO"
+	VerifyingKey      HashID = "VK"
+	EphemeralPK       HashID = "EP"
```
Comment (winder) [inline protocol/hash.go:59]: nit: these should be sorted according to the comment above

File: crypto/merklekeystore/keystore_test.go [hunk 11]
```diff
@@ -0,0 +1,212 @@
...
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package merklekeystore
+
+import (
+	"crypto/rand"
+	"github.com/algorand/go-algorand/crypto"
+	"github.com/stretchr/testify/require"
+	"math"
+	"testing"
```
Comment (winder) [inline crypto/merklekeystore/keystore_test.go:24]: nit: imports aren't following our style conventions.

```suggestion
	"crypto/rand"
	"math"
	"testing"

	"github.com/stretchr/testify/require"

	"github.com/algorand/go-algorand/crypto"
```

File: crypto/merklekeystore/keystore.go [hunk 12]
```diff
@@ -0,0 +1,179 @@
...
+
+import (
+	"fmt"
+	"github.com/algorand/go-algorand/crypto"
+	"github.com/algorand/go-algorand/crypto/merklearray"
+	"github.com/algorand/go-algorand/protocol"
+)
+
+type (
+	//EphemeralKeys represent the possible keys inside the keystore.
```
Comment (algoidan) [inline crypto/merklekeystore/keystore.go:27]: try to mention that you have multiple keys which will be used for several rounds

**Review by algoidan — COMMENTED** [review]
great job!

**Pull request merged**